src/aof.c

```diff
@@ -1083,10 +1083,21 @@ void flushAppendOnlyFile(int force) {
          * stop write commands before fsync called in one second,
          * the data in page cache cannot be flushed in time. */
         if (server.aof_fsync == AOF_FSYNC_EVERYSEC &&
-            server.aof_fsync_offset != server.aof_current_size &&
+            (server.aof_fsync_dirty ||
+            (!server.aof_fsync_dirty && server.aof_fsync_offset != server.aof_current_size)) &&
             server.unixtime > server.aof_last_fsync &&
             !(sync_in_progress = aofFsyncInProgress())) {
             goto try_fsync;
+
+        /* Check if we need to do fsync even the aof buffer is empty,
+         * the reason is described in the previous AOF_FSYNC_EVERYSEC if,
+         * and AOF_FSYNC_ALWAYS is also checked here to prevent users from
+         * changing the aof_fsync from everysec to always.  */
```

Comment:
```suggestion
         * the reason is described in the previous AOF_FSYNC_EVERYSEC block,
         * and AOF_FSYNC_ALWAYS is also checked here to handle a case where
         * aof_fsync i changed from everysec to always.  */
```

maybe a better solution could be to just explicitly fsync in updateAppendFsync?

Reply:
I thought about it, but I ended up putting it here
- These checks can be put with AOF_ FSYNC_ EVERYSEC check block.
- The fsync related code is in the flushAppendOnlyFile

or you mean we do a flushAppendOnlyFile(1) in updateAppendFsync?

Reply:
i thought about doing `redis_fsync(server.aof_fd)` in `updateAppendFsync()`.
i.e. it'll run once on any configuration change that is done at runtime (very rare act to normalize things),
instead of an extra condition in flushAppendOnlyFile (constantly called by beforeSleep)

Reply:
yes, that makes sense. let me know if it needs to be changed like this, i am ok with it 
though i am very worried that I need to copy this piece of code...
```c
    if (server.aof_fsync == AOF_FSYNC_ALWAYS) {
        /* redis_fsync is defined as fdatasync() for Linux in order to avoid
         * flushing metadata. */
        latencyStartMonitor(latency);
        /* Let's try to get this data on the disk. To guarantee data safe when
         * the AOF fsync policy is 'always', we should exit if failed to fsync
         * AOF (see comment next to the exit(1) after write error above). */
        if (redis_fsync(server.aof_fd) == -1) {
            serverLog(LL_WARNING,"Can't persist AOF for fsync error when the "
              "AOF fsync policy is 'always': %s. Exiting...", strerror(errno));
            exit(1);
        }
        latencyEndMonitor(latency);
        latencyAddSampleIfNeeded("aof-fsync-always",latency);
        server.aof_fsync_dirty = 0;
        server.aof_fsync_offset = server.aof_current_size;
        server.aof_last_fsync = server.unixtime;
        atomicSet(server.fsynced_reploff_pending, server.master_repl_offset);
```

Reply:
you're right. i didn't consider all the extra state we need (or want) to do besides fsync.
i would not like to copy / clone that to config.c
the other option is to extract these to a function so that it can be called from both config.c and aof.c.
your way may be better for the blame log, and maybe extracting them to a separate function would be better for future generations.
i'm leaning towards keeping your way, let me know what you think.

Reply:
i'm also leaning towards keeping my way.

I'm wondering if this plan is ok: https://github.com/redis/redis/pull/11973#issuecomment-1486711096
do you have some thoughts?

Reply:
> I still don't like the "dirty" flag, `aof_fsync_offset` can work well if we compare it with `aof_last_incr_size` instead of `aof_current_size`.

well, first, that comment is also applicable to EVERYSEC, so let's first put the discussion about ALWAYS behind us, and agree that we keep your approach.

secondly, when reading that comment i wonder again if there was indeed a bug in 6.2 (before multi-part-aof), and conclude (as already mentioned in the top comment), that it was only when switching from `everysec` to `always`.
if the bug is just when switching (not a steady state `always`), i'm having a second thought if we should backport it to 6.2 or not. please correct me if i'm wrong.

lastly regarding @soloestoy's suggestion, could it be that it'll still get the wrong conclusion if on a rare case, by the time we get to beforeSleep, the new INCR file is the same size as the previous INCR file?
e.g. either there are a lot of commands in a single event loop, or maybe if the previous INCR file only contain one command, and then someone did a BGREWRITEAOF and a single command (both processed inside the same event loop)? maybe the dirty flag is safer...

Reply:
> could it be that it'll still get the wrong conclusion if on a rare case, by the time we get to beforeSleep, the new INCR file is the same size as the previous INCR file?

`aof_fsync_offset` should be reset to 0 when open a new INCR file, so it is safe.

Reply:
ok, and then any write to the AOF file will "dirtify" it.
so i think that's ok, and better than adding another variable.
considering we'll have test for these cases now, i think it should be safe to adopt that simpler design with less conditions in the `if` and less variables.

Reply:
> secondly, when reading that comment i wonder again if there was indeed a bug in 6.2 (before multi-part-aof), and conclude (as already mentioned in the top comment), that it was only when switching from everysec to always.
if the bug is just when switching (not a steady state always), i'm having a second thought if we should backport it to 6.2 or not. please correct me if i'm wrong.

yep, indeed, it will only happen in everysec -> always. the case is: in everysec, we write but skip the fsync (due to second), and then stop write and switch to always. At this time, the skipped fsync will still be un-fsync.

Reply:
so i would say that old versions have bugs, and we don't always have to rush to fix them.
in this case the bug is a rather rare scenario (switching fsync configuration at runtime), and the impact isn't a guranteed crash or data corruption, it's just a temporary fsync safety compromise.

so if the backport is gonna be complicated (more than just fix one `if` condition and add a test), i think we can drop it.
on the other hand, maybe before multi-part AOF, there's no need for anything other than one line fix for the `if` condition. is that right?

in that case, i don't think we wanna keep the plan of splitting this PR into two so one can be cherry picked.
instead, we'll create some 6.2 based branch with that fix, and handle it manually when the day comes.

Reply:
> on the other hand, maybe before multi-part AOF, there's no need for anything other than one line fix for the if condition. is that right?

right

so, two options:
1. two PR: one is the if condition, `server.aof_fsync == AOF_FSYNC_ALWAYS && server.aof_fsync_offset != server.aof_current_size`. It is easy to merge and cherry-pick to old version.
The other one is this one, merge into unstable and cherry-pick to 7.0

2. Just merge this one into unstable, and cherry-pick to 7.0.
if we want to fix the old version issue, create another PR based on 6.2 branch

i am ok with both options

Reply:
yeah, but the condition in option 1 being merged to unstable isn't suitable for unstable (won't work there properly).
and it does need to come with a test...
i have a feeling we spend too much time on a rare, low impact 6.2 bug, let's just forget it.

Reply:
> yeah, but the condition in option 1 being merged to unstable isn't suitable for unstable (won't work there properly).

It actually works, it's just like the everysec above. (has nothing to do with MP-AOF or WAITAOF, the change aof_last_incr_size is for MP-AOF). 

yean, speed too much time on it. let's forget it.

Reply:
but in unstable it needs to use `aof_last_incr_size`, not `aof_current_size`.
if you make a PR that should be merged first and then backported to 6.2, it also needs to come with portion of the tests.

Reply:
ohh, i got your point. it can actually be divided into two steps:
- change always like everysec (using aof_fsync_offset and aof_current_size). what I said works is this step
- then change aof_fsync_offset to aof_last_incr_fsync_offset, change aof_current_size to aof_last_incr_size. this is for MP-AOF

Reply:
let's drop it.. it doesn't worth the effort.